server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,217 @@
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(mail.getMessageSize());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.MESSAGE_SIZE, e);
+        }
+    }
+
+    private static Optional<String> fetchTextBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.TEXT_BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function(extractor::extract).sneakyThrow())
+                    .flatMap(MessageContent::getTextBody);
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.TEXT_BODY, e);
+        }
+    }
+
+    private static Optional<String> fetchHtmlBody(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HTML_BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function(extractor::extract).sneakyThrow())
+                    .flatMap(MessageContent::getHtmlBody);
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.HTML_BODY, e);
+        }
+    }
+
+    private static Message convertMessage(MimeMessage message) throws IOException, MessagingException {
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        message.writeTo(rawMessage);
+        return Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(new ByteArrayInputStream(rawMessage.toByteArray()))
+                .build();
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(Set<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::extractHeaders).sneakyThrow());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.HEADERS, e);
+        }
+    }
+
+    private static HeadersDto extractHeaders(MimeMessage message) throws MessagingException {
+        return new HeadersDto(Collections
+            .list(message.getAllHeaders())
```

Comment:
What about header unscrambling in HeaderDTO? 

`MimeUtil.unscrambleHeaderValue(headerValue)`

Overwise we will encounter encoding issues...

Reply:
Headers can have encoding?

Do you mean I have to extract this method in MimeUtils?